userchroot.c

```diff
@@ -682,6 +682,46 @@ int main(int argc, char* argv[], char* envp[]) {
         }
       }
 
+      // Cleanup the /dev/shm directory created by child_fn
+      int shm_path_len = strlen(final_path) + strlen("/dev/shm") + 1;
+      char* shm_path = malloc(shm_path_len);
+      if (shm_path == NULL) {
+        fprintf(stderr,"Failed to allocate memory. Aborting.\n");
+        exit(ERR_EXIT_CODE);
+      }
+      rc = snprintf(shm_path, shm_path_len, "%s/%s", final_path, "dev/shm");
+      if (rc <= 0) {
+        fprintf(stderr,"Failed to assemble path. Aborting.\n");
+        exit(ERR_EXIT_CODE);
+      }
+      rc = rmdir(shm_path);
```

Comment:
Shall we make this conditional to remove `/dev/shm` (and `/proc`) only if we actually created it (i.e., if `mkdir()` did not fail with `EEXIST`)?

For chroot directories that are used only by a single userchroot session at a time, this would be the cleanest approach. It might not be ideal for chroot directories that are used by multiple concurrent userchroot sessions, though.

Reply:
Maybe we could make it conditional on `st_uid`. If setuid root `userchroot` creates the directory, I'd expect `st_uid` to be zero. If the chroot owner creates the directory, it will be non-zero. Calling `rmdir()` only if `st_uid` is zero should work nicely in all cases if I'm not forgetting anything.

Moving `mkdir()` and `rmdir()` to `--install-devices` and `--uninstall-devices`, respectively, could indeed make sense. With the `st_uid` check it should work either way, though.

Reply:
> Maybe we could make it conditional on `st_uid`

I don't think I'm following. If you're running two concurrent `userchroot` processes against the same chroot, wouldn't one of them exiting still remove those directories for the other? The created `/proc` and `/dev/shm` directories are `st_uid` owned even if there are other ongoing `userchroot` invocations.  Or are we talking about two totally different things?

Reply:
If `/dev/shm` is still mounted by another `userchroot` session, `rmdir()` should simply fail with `EBUSY`. We could ignore that and the directory would get cleaned up when the last `userchroot` process terminates.

There is a race condition, though. One `userchroot` process would be able to remove the directory while another `userchroot` process is between `mkdir()` and `mount()`. Moving `mkdir()` and `rmdir()` to `--install-devices` and `--uninstall-devices`, respectively, fixes that race condition, so I think we should indeed do that (maybe first merge #17).

The point of the `st_uid` conditional is to avoid removing directories that we didn't create. Installing devices, running a command via `userchroot`, and then uninstalling devices should ideally result in an unmodified chroot directory, in my opinion (except for intentional changes performed by the command). And thus, I still think we should add this conditional.

Reply:
> If /dev/shm is still mounted by another userchroot session, rmdir() should simply fail with `EBUSY`

Interestingly, this hasn't happened to me in my testing. What I've seen happen when testing these changes locally is the first `userchroot` process exits successfully and both `/proc` and `/dev/shm` simply disappear from the view of the second, still running `userchroot` process. I was testing by using `userchroot` to run two bash processes, if that matters.

> The point of the `st_uid` conditional is to avoid removing directories that we didn't create

Makes sense. I'll incorporate that check into a new version that calls `mkdir()` during `--install-devices` and `rmdir()` in `--uninstall-devices`

Reply:
Ah, for security reasons (DoS) `EBUSY` is returned only if the mount point is used in the same mount namespace. If it's mounted in another namespace, it's implicitly unmounted and `rmdir()` succeeds.

> Makes sense. I'll incorporate that check into a new version that calls `mkdir()` during `--install-devices` and `rmdir()` in `--uninstall-devices`

Sounds good, thanks.

Reply:
Pushed up a new version that's ready to be looked at.